backend/src/main/java/ai/verta/modeldb/experimentRun/FutureExperimentRunDAO.java

```diff
@@ -365,4 +418,362 @@ public FutureExperimentRunDAO(Executor executor, FutureJdbi jdbi, UAC uac) {
+        // newExperimentRun.getArtifactsList()), executor)
+        // TODO .thenCompose(handle -> datasetHandler.logDatasets(newExperimentRun.getId(),
+        // newExperimentRun.getDatasetsList()), executor)
+        .thenCompose(
+            handle ->
+                observationHandler.logObservations(
+                    newExperimentRun.getId(), newExperimentRun.getObservationsList(), now),
+            executor)
+        // TODO .thenCompose(handle -> featureHandler.logFeatures(newExperimentRun.getId(),
+        // newExperimentRun.getFeaturesList()), executor)
+        // TODO .thenCompose(handle -> addCodeVersionSnapShot(), executor)
+        // TODO .thenCompose(handle -> versioned_inputs, executor)
+
+        .thenCompose(unused -> InternalFuture.completedInternalFuture(newExperimentRun), executor);
+  }
+
+  private EnvironmentBlob sortPythonEnvironmentBlob(EnvironmentBlob environmentBlob) {
+    EnvironmentBlob.Builder builder = environmentBlob.toBuilder();
+    if (builder.hasPython()) {
+      PythonEnvironmentBlob.Builder pythonEnvironmentBlobBuilder = builder.getPython().toBuilder();
+
+      // Compare requirementEnvironmentBlobs
+      List<PythonRequirementEnvironmentBlob> requirementEnvironmentBlobs =
+          new ArrayList<>(pythonEnvironmentBlobBuilder.getRequirementsList());
+      requirementEnvironmentBlobs.sort(
+          Comparator.comparing(PythonRequirementEnvironmentBlob::getLibrary));
+      pythonEnvironmentBlobBuilder
+          .clearRequirements()
+          .addAllRequirements(requirementEnvironmentBlobs);
+
+      // Compare
+      List<PythonRequirementEnvironmentBlob> constraintsBlobs =
+          new ArrayList<>(pythonEnvironmentBlobBuilder.getConstraintsList());
+      constraintsBlobs.sort(Comparator.comparing(PythonRequirementEnvironmentBlob::getLibrary));
+      pythonEnvironmentBlobBuilder.clearConstraints().addAllConstraints(constraintsBlobs);
+
+      builder.setPython(pythonEnvironmentBlobBuilder.build());
+    }
+    return builder.build();
+  }
+
+  private InternalFuture<ExperimentRun> checkIfEntityAlreadyExists(
+      ExperimentRun experimentRun, Boolean isInsert) {
+    return jdbi.useHandle(
+            handle -> {
+              try {
+                String queryStr;
+                if (isInsert) {
+                  queryStr =
+                      "SELECT id FROM experiment_run ere WHERE "
+                          + " ere."
+                          + ModelDBConstants.NAME
+                          + " = :experimentRunName "
+                          + " AND ere."
+                          + ModelDBConstants.PROJECT_ID
+                          + " = :projectId "
+                          + " AND ere."
+                          + ModelDBConstants.EXPERIMENT_ID
+                          + " = :experimentId "
+                          + " AND ere."
+                          + ModelDBConstants.DELETED
```

Comment:
Avoid this constant patching with constants. It really hurts readability. Just type the field name instead. It's really bad practice to depend on a kitchen sink central list of constants.

Reply:
Removed constant fields from query and added fields directly in query  string